site/topics/lists/lists.rst

```diff
@@ -112,13 +112,173 @@ Mutability
+
+.. code-block:: python
+    :linenos:
+
+    collection_of_things = ["Hello", 10, True, 100.001]
+
+    for thing in collection_of_things:
+      print(thing)
+
+    # Results in
+    #   Hello
+    #   10
+    #   True
+    #   100.001
+
+
+* Iterating over a *collection of things* is very common
+* Expect to start using ``for`` loops like this a lot
+* In fact, this was used in assignment 1 multiple times
+
+    * Iterating over the contents of the file being read
+    * Iterating over the list of ``(latitude, longitude)`` pairs
+
+
+Range
+-----
+
+* ``range`` is a very handy function that we often use with ``for`` loops
+* It provides an easy way to loop over a specific range of numbers
+
+.. code-block:: python
+    :linenos:
+
+    for i in range(5):
+      print(i)
+
+    # Results in
+    #   0
+    #   1
+    #   2
+    #   3
+    #   4
+
+
+* Notice that the integer ``5`` was specified in the ``range`` function, and the loop ran a total of ``5`` times
+
+    * But because of ``0`` based indexing, the the number ``5`` is not actually included
+
+* A big reason we like to use ``for`` loops this way is because the syntax is so simple and clean
+* The above functionality can be achieved with a ``while`` loop, but the code needed is a little more cumbersome
+
+.. code-block:: python
+    :linenos:
+
+    i = 0
+    while i < 5:
+      print(i)
+      i += 1
+
+    # Results in
+    #   0
+    #   1
+    #   2
+    #   3
+    #   4
+
+
+.. admonition:: Activity
+    :class: activity
+
+    Write a function ``beer_on_wall`` that prints out ``"n bottles of beer on the wall"`` for all ``n`` from ``99`` down
+    to ``0``. This function must use a ``for`` loop with the ``range`` function.
+
+    The difficulty here is the need to count backwards.
+
+    `Perhaps a read of the documentation can help. <https://docs.python.org/3/library/stdtypes.html#typesseq-range>`_
+
+    .. raw:: html
+
+        <iframe width="560" height="315" src="https://www.youtube.com/embed/0AuMKIMiPMo" frameborder="0" allowfullscreen></iframe>
+
+
+Looping Over Indices and Enumerate
+----------------------------------
+
+* Sometimes it is important to know the indices of each element being printed out
+* For example, if the goal is to do a linear search to find the index of a given element, the following would be problematic
+
+.. code-block:: python
+    :linenos:
+
+    def index_of_uhoh(needle, haystack):
+        for element in haystack:
+            if element == needle:
+                return ???? # How do I find the index?
+        return -1
+
+
+* Using the ``range`` function would help in this situation as we can use it to loop over each index in ``haystack``
+* The only catch is that we need to actually index ``haystack``
+
+.. code-block:: python
+    :linenos:
+
+    def index_of(needle, haystack):
+        haystack_length = len(haystack)
+        for i in range(haystack_length):
+            if haystack[i] == needle:
```

Comment:
[enumerate](https://docs.python.org/3/library/functions.html#enumerate) is specifically for this case.